lib/queuePopulator/LogReader.js

```diff
@@ -292,38 +305,67 @@ class LogReader {
         });
         logRes.log.on('end', () => {
             this.log.debug('ending record stream');
+            this._endEntryBatch();
             return done();
         });
         return undefined;
     }
 
+    _setupProducer(topic, done) {
+        if (this._producers[topic] !== undefined) {
+            return process.nextTick(done);
+        }
+        const producer = new BackbeatProducer({
+            zookeeper: { connectionString: this.zkConfig.connectionString },
+            topic,
+        });
+        producer.once('error', done);
+        producer.once('ready', () => {
+            producer.removeAllListeners('error');
+            producer.on('error', err => {
+                this.log.error('error from backbeat producer',
+                               { topic, error: err });
+            });
+            this._producers[topic] = producer;
+            done();
+        });
+        return undefined;
+    }
+
     _processPublishEntries(batchState, done) {
         const { entriesToPublish, logRes, logStats } = batchState;
 
-        if (entriesToPublish.length === 0) {
-            if (logRes.info.start !== null) {
-                batchState.nextLogOffset =
-                    logRes.info.start + logStats.nbLogRecordsRead;
-            }
+        if (logRes.info.start === null) {
             return done();
         }
-        return this.producer.send(entriesToPublish, err => {
-            if (err) {
-                this.log.error(
-                    'error publishing entries from log',
-                    { method: 'LogReader._processPublishEntries',
-                        error: err });
-                return done(err);
+        batchState.nextLogOffset =
+            logRes.info.start + logStats.nbLogRecordsRead;
+
+        return async.each(Object.keys(entriesToPublish), (topic, done) => {
+            const topicEntries = entriesToPublish[topic];
+            if (topicEntries.length === 0) {
+                return done();
             }
-            batchState.nextLogOffset =
-                logRes.info.start + logStats.nbLogRecordsRead;
-            this.log.debug(
-                'entries published successfully',
-                { method: 'LogReader._processPublishEntries',
-                    entryCount: entriesToPublish.length,
-                    logOffset: batchState.nextLogOffset });
-            return done();
-        });
+            return async.series([
+                done => this._setupProducer(topic, done),
```

Comment:
The inner `done` is guaranteed to be used since it's the closest scope, I personally don't mind though if you guys think it's confusing I can rework to avoid duplicate callback names.